From 5700feeb83b8b7d80b1589cd89aeeff16fb67b31 Mon Sep 17 00:00:00 2001
From: Harshitha Prem <quic_hprem@quicinc.com>
Date: Thu, 21 Nov 2024 12:04:00 +0530
Subject: [PATCH] QSDK:CP:mac80211.sh: Add support to bringin features with
 openwrt upgrade

Add support to bring in the below features
	1) Discard AWGN event
	2) 80 MHz, 160 MHz and 320 MHz EHT ULMUMIMO support
	3) CCFS
	4) RU puncture OFDMA
	5) RU puncture DFS
	6) RU puncture bitmap
	7) RU puncture ACS Threshold
	8) Disable CSA DFS
	9) Radio id support for wds scenario.
	10) Add EHT with 320 Mhz support for repeater-AP.
	11) ML reconfigure from wifi command
	12) PPE VP
	13) BSS color calculation
	14) FQ limit configuration
	15) single wiphy capability string
	16) HE configurations
	17) EHT configurations
	18) smp affinity and performance scripts

Add changes for generating default interface names:
	For non-single wiphy architecture: phy00-ap0, phy00-ap1
	For single wiphy architecture: phy00-2g-ap0, phy00-5g-ap0
	for MLD vaps, phy00-mld1 , where mld1 is the mld name

Also, add changes to set freq_list for mesh and sta config

Co-developed-by: Nagarajan Maran <quic_nmaran@quicinc.com>
Change-Id: I2299484caf0da905d2729d0b9ca7df097baf374f
Signed-off-by: Nagarajan Maran <quic_nmaran@quicinc.com>
Signed-off-by: Harshitha Prem <quic_hprem@quicinc.com>
---
 .../files/lib/netifd/wireless/mac80211.sh     | 788 +++++++++++++++---
 1 file changed, 695 insertions(+), 93 deletions(-)

diff --git a/package/network/config/wifi-scripts/files/lib/netifd/wireless/mac80211.sh b/package/network/config/wifi-scripts/files/lib/netifd/wireless/mac80211.sh
index 9998e26ec131..7b5712060e8d 100755
--- a/package/network/config/wifi-scripts/files/lib/netifd/wireless/mac80211.sh
+++ b/package/network/config/wifi-scripts/files/lib/netifd/wireless/mac80211.sh
@@ -3,7 +3,11 @@
 . /lib/netifd/hostapd.sh
 . /lib/functions/system.sh
 
-init_wireless_driver "$@"
+mlo_add_flag=0
+[ -f /tmp/mlo_support.txt ] && mlo_add_flag=$(cat /tmp/mlo_support.txt)
+if [ $mlo_add_flag -eq 0 ]; then
+	init_wireless_driver "$@"
+fi
 
 MP_CONFIG_INT="mesh_retry_timeout mesh_confirm_timeout mesh_holding_timeout mesh_max_peer_links
 	       mesh_max_retries mesh_ttl mesh_element_ttl mesh_hwmp_max_preq_retries
@@ -15,6 +19,67 @@ MP_CONFIG_INT="mesh_retry_timeout mesh_confirm_timeout mesh_holding_timeout mesh
 MP_CONFIG_BOOL="mesh_auto_open_plinks mesh_fwding"
 MP_CONFIG_STRING="mesh_power_mode"
 
+is_wiphy_multi_radio=0
+
+#update he params
+bss_color=
+enable_color=
+
+#ACS DFS
+acs_exclude_dfs=
+
+min_tx_power=
+noscan=
+ht_coex=
+rx_stbc=
+ldpc=
+greenfield=
+short_gi_20=
+short_gi_40=
+tx_stbc=
+max_amsdu=
+dsss_cck_40=
+background_radar=
+rxldpc=
+short_gi_80=
+short_gi_160=
+tx_stbc_2by1=
+su_beamformer=
+su_beamformee=
+mu_beamformer=
+vht_txop_ps=
+htc_vht=
+rx_antenna_pattern=
+tx_antenna_pattern=
+vht160=
+vht_max_mpdu=
+vht_link_adapt=
+vht_max_a_mpdu_len_exp=
+disable_eml_cap=
+eht_ulmumimo_80mhz=
+eht_ulmumimo_160mhz=
+eht_ulmumimo_320mhz=
+ru_punct_bitmap=
+ru_punct_ofdma=
+ru_punct_acs_threshold=
+use_ru_puncture_dfs=
+he_su_beamformer=
+he_su_beamformee=
+he_mu_beamformer=
+he_spr_psr_enabled=
+he_twt_required=
+he_ul_mumimo=
+he_bss_color_enabled=
+he_spr_non_srg_obss_pd_max_offset=
+disable_csa_dfs=
+discard_6g_awgn_event=
+wds=
+wds_bridge=
+start_disabled=
+dtim_period=
+max_listen_int=
+he_6ghz_reg_pwr_type=
+
 wdev_tool() {
 	ucode /usr/share/hostap/wdev.uc "$@"
 }
@@ -28,9 +93,9 @@ drv_mac80211_init_device_config() {
 
 	config_add_string path phy 'macaddr:macaddr'
 	config_add_string tx_burst
-	config_add_string distance
+	config_add_string distance band
 	config_add_int radio beacon_int chanbw frag rts
-	config_add_int rxantenna txantenna txpower min_tx_power
+	config_add_int rxantenna txantenna txpower min_tx_power antenna_gain
 	config_add_int num_global_macaddr multiple_bssid
 	config_add_boolean noscan ht_coex acs_exclude_dfs background_radar
 	config_add_array ht_capab
@@ -55,7 +120,9 @@ drv_mac80211_init_device_config() {
 		he_spr_sr_control \
 		he_spr_psr_enabled \
 		he_bss_color_enabled \
-		he_twt_required
+		he_twt_required \
+		ru_punct_ofdma \
+		use_ru_puncture_dfs
 	config_add_int \
 		beamformer_antennas \
 		beamformee_antennas \
@@ -66,27 +133,45 @@ drv_mac80211_init_device_config() {
 		rx_stbc \
 		tx_stbc \
 		he_bss_color \
-		he_spr_non_srg_obss_pd_max_offset
+		he_spr_non_srg_obss_pd_max_offset \
+		he_spr_sr_control \
+		he_ul_mumimo \
+		eht_ulmumimo_160mhz \
+		eht_ulmumimo_320mhz \
+		ru_punct_bitmap \
+		ru_punct_acs_threshold \
+		ccfs \
+		multiple_bssid \
+		mbssid_group_size \
+		he_6ghz_reg_pwr_type
 	config_add_boolean \
 		ldpc \
 		greenfield \
 		short_gi_20 \
 		short_gi_40 \
 		max_amsdu \
-		dsss_cck_40
+		dsss_cck_40 \
+		disable_eml_cap \
+		disable_csa_dfs \
+		discard_6g_awgn_event
 }
 
 drv_mac80211_init_iface_config() {
 	hostapd_common_add_bss_config
 
-	config_add_string 'macaddr:macaddr' ifname
+	config_add_string 'macaddr:macaddr' ifname mld
 
 	config_add_boolean wds powersave enable
-	config_add_string wds_bridge
+	config_add_string wds_bridge mld
 	config_add_int maxassoc
 	config_add_int max_listen_int
 	config_add_int dtim_period
 	config_add_int start_disabled
+	config_add_int ieee80211w
+	config_add_int beacon_prot
+	config_add_int unsol_bcast_presp
+	config_add_int fils_discovery
+	config_add_string ppe_vp
 
 	# mesh
 	config_add_string mesh_id
@@ -131,8 +216,43 @@ mac80211_add_he_capabilities() {
 	IFS="$oifs"
 }
 
+get_sta_freq_list() {
+	local phy=$1
+	local radio_id=$2
+	local band_name=$3
+
+	[ "$radio_id" = "-1" ] && return ""
+
+	freq_range=$(iw $phy info | grep -A 1 "Idx $radio_id:" | grep "Frequency Range:" | awk '{print $3, $6}')
+	set -- $freq_range
+	freq1=$1
+	freq2=$2
+
+        case "$band_name" in
+                2g) band="1:";;
+                5gl|5gh|5g) band="2:";;
+                60g) band="3:";;
+                6g|6gl|6gh) band="4:";;
+        esac
+
+	iw "$phy" info | awk -v band="$band" -v min="$freq1" -v max="$freq2" '
+
+$1 ~ /Band/ {
+        band_match = band == $2
+}
+
+band_match && $3 == "MHz" {
+        freq = $2
+         if (freq >= min && freq <= max) {
+                print int($2)
+        }
+}
+' | tr '\n' ' '
+}
+
 mac80211_hostapd_setup_base() {
 	local phy="$1"
+	local sedString=
 
 	json_select config
 
@@ -142,15 +262,24 @@ mac80211_hostapd_setup_base() {
 	[ -n "$acs_exclude_dfs" ] && [ "$acs_exclude_dfs" -gt 0 ] &&
 		append base_cfg "acs_exclude_dfs=1" "$N"
 
-	json_get_vars noscan ht_coex min_tx_power:0 tx_burst
+	json_get_vars noscan ht_coex min_tx_power:0 tx_burst disable_csa_dfs use_ru_puncture_dfs
 	json_get_values ht_capab_list ht_capab
 	json_get_values channel_list channels
+	json_get_vars disable_eml_cap discard_6g_awgn_event ccfs
 
 	[ "$auto_channel" = 0 ] && [ -z "$channel_list" ] && \
 		channel_list="$channel"
 
 	[ "$min_tx_power" -gt 0 ] && append base_cfg "min_tx_power=$min_tx_power" "$N"
 
+	if [ "$band" = "2g" ]; then
+		sedString="iw phy ${phy} info | awk  '/Band 1/{ f = 1; next } /Band /{ f = 0 } f'"
+	elif [ "$band" = "5g" ]; then
+		sedString="iw phy ${phy} info | awk  '/Band 2/{ f = 1; next } /Band /{ f = 0 } f'"
+	elif [ "$band" = "6g" ]; then
+		sedString="iw phy ${phy} info | awk  '/Band 4/{ f = 1; next } /Band /{ f = 0 } f'"
+	fi
+
 	set_default noscan 0
 
 	[ "$noscan" -gt 0 ] && hostapd_noscan=1
@@ -164,7 +293,7 @@ mac80211_hostapd_setup_base() {
 		ht_capab=
 		case "$htmode" in
 			VHT20|HT20|HE20|EHT20) ;;
-			HT40*|VHT40|VHT80|VHT160|HE40|HE80|HE160|EHT40|EHT80|EHT160)
+			HT40*|VHT40|VHT80|VHT160|HE40|HE80|HE160|EHT40|EHT80|EHT160|EHT320)
 				case "$hwmode" in
 					a)
 						case "$(( (($channel / 4) + $chan_ofs) % 2 ))" in
@@ -208,7 +337,7 @@ mac80211_hostapd_setup_base() {
 				dsss_cck_40:1
 
 			ht_cap_mask=0
-			for cap in $(iw phy "$phy" info | grep 'Capabilities:' | cut -d: -f2); do
+			for cap in $(eval $sedString | grep -E '^\s*Capabilities:' | cut -d: -f2); do
 				ht_cap_mask="$(($ht_cap_mask | $cap))"
 			done
 
@@ -237,17 +366,31 @@ mac80211_hostapd_setup_base() {
 	enable_ac=0
 	vht_oper_chwidth=0
 	vht_center_seg0=
+	eht_oper_chwidth=0
+	eht_center_seg0=
 
 	idx="$channel"
 	case "$htmode" in
-		VHT20|HE20|EHT20) enable_ac=1;;
+		VHT20|HE20|EHT20)
+			enable_ac=1
+			if [ "$hwmode" = "a" ]; then
+				vht_oper_chwidth=0
+				vht_center_seg0=$idx
+				eht_center_seg0=$idx
+			fi
+		;;
 		VHT40|HE40|EHT40)
 			case "$(( (($channel / 4) + $chan_ofs) % 2 ))" in
 				1) idx=$(($channel + 2));;
 				0) idx=$(($channel - 2));;
 			esac
 			enable_ac=1
+			vht_oper_chwidth=0
 			vht_center_seg0=$idx
+			if [ "$hwmode" = "a" ]; then
+				vht_oper_chwidth=0
+				vht_center_seg0=$idx
+			fi
 		;;
 		VHT80|HE80|EHT80)
 			case "$(( (($channel / 4) + $chan_ofs) % 4 ))" in
@@ -289,20 +432,39 @@ mac80211_hostapd_setup_base() {
 		[ "$background_radar" -eq 1 ] && append base_cfg "enable_background_radar=1" "$N"
 	}
 
-	eht_oper_chwidth=$vht_oper_chwidth
-	eht_center_seg0=$vht_center_seg0
+	if [ "$htmode" = "EHT320" ]; then
+		eht_oper_chwidth=9
+		if [ "$freq" -ge 5500 ] && [ "$freq" -le 5730 ]; then
+			eht_center_seg0=130
+		else
+			eht_center_seg0=$vht_center_seg0
+		fi
+	else
+		eht_oper_chwidth=$vht_oper_chwidth
+		eht_center_seg0=$vht_center_seg0
+	fi
 
 	[ "$band" = "6g" ] && {
 		op_class=
 		case "$htmode" in
-			HE20|EHT20) op_class=131;;
+			HE20|EHT20)
+				if [ "$freq" == "5935" ]; then
+					op_class=136
+				else
+					op_class=131
+				fi
+			;;
 			EHT320)
-				case "$channel" in
-					1|5|9|13|17|21|25|29|33|37|41|45|49|53|57|61) idx=31;;
-					65|69|73|77|81|85|89|93|97|101|105|109|113|117|121|125) idx=95;;
-					129|133|137|141|145|149|153|157|161|165|169|173|177|181|185|189) idx=159;;
-					193|197|201|205|209|213|217|221) idx=191;;
-				esac
+				if [ -n "$ccfs" ] && [ "$ccfs" -gt 0 ]; then
+					idx="$ccfs"
+				elif [ -z "$ccfs" ] || [ "$ccfs" -eq "0" ]; then
+					case "$channel" in
+						1|5|9|13|17|21|25|29|33|37|41|45|49|53|57|61) idx=31;;
+						65|69|73|77|81|85|89|93|97|101|105|109|113|117|121|125) idx=95;;
+						129|133|137|141|145|149|153|157|161|165|169|173|177|181|185|189) idx=159;;
+						193|197|201|205|209|213|217|221) idx=191;;
+					esac
+				fi
 
 				op_class=137
 				eht_center_seg0=$idx
@@ -312,6 +474,7 @@ mac80211_hostapd_setup_base() {
 		esac
 		[ -n "$op_class" ] && append base_cfg "op_class=$op_class" "$N"
 	}
+
 	[ "$hwmode" = "a" ] || enable_ac=0
 	[ "$band" = "6g" ] && enable_ac=0
 
@@ -340,7 +503,7 @@ mac80211_hostapd_setup_base() {
 		set_default tx_burst 2.0
 		append base_cfg "ieee80211ac=1" "$N"
 		vht_cap=0
-		for cap in $(iw phy "$phy" info | awk -F "[()]" '/VHT Capabilities/ { print $2 }'); do
+		for cap in $(eval $sedString | awk -F "[()]" '/VHT Capabilities/ { print $2 }'); do
 			vht_cap="$(($vht_cap | $cap))"
 		done
 
@@ -388,12 +551,18 @@ mac80211_hostapd_setup_base() {
 
 		# supported Channel widths
 		vht160_hw=0
-		[ "$(($vht_cap & 12))" -eq 4 -a 1 -le "$vht160" ] && \
-			vht160_hw=1
-		[ "$(($vht_cap & 12))" -eq 8 -a 2 -le "$vht160" ] && \
-			vht160_hw=2
-		[ "$vht160_hw" = 1 ] && vht_capab="$vht_capab[VHT160]"
-		[ "$vht160_hw" = 2 ] && vht_capab="$vht_capab[VHT160-80PLUS80]"
+		case "$htmode" in
+			VHT160|HE160|EHT160|EHT320)
+				([ "$(($vht_cap & 12))" -eq 4 ] && [ 1 -le "$vht160" ]) && \
+				vht160_hw=1
+				[ "$vht160_hw" = 1 ] && vht_capab="$vht_capab[VHT160]"
+				;;
+			VHT80+80|HE80+80)
+				([ "$(($vht_cap & 12))" -eq 8 ] && [ 2 -le "$vht160" ]) && \
+				vht160_hw=2
+				[ "$vht160_hw" = 2 ] && vht_capab="$vht_capab[VHT160-80PLUS80]"
+				;;
+		esac
 
 		# maximum MPDU length
 		vht_max_mpdu_hw=3895
@@ -439,7 +608,9 @@ mac80211_hostapd_setup_base() {
 	enable_be=0
 	case "$htmode" in
 		HE*) enable_ax=1 ;;
-		EHT*) enable_ax=1; enable_be=1 ;;
+		EHT*) enable_ax=1; enable_be=1
+		      [ -n "$disable_eml_cap" ] && append base_cfg "disable_eml_cap=$disable_eml_cap" "$N"
+		;;
 	esac
 
 	if [ "$enable_ax" != "0" ]; then
@@ -452,11 +623,22 @@ mac80211_hostapd_setup_base() {
 			he_spr_psr_enabled:0 \
 			he_spr_non_srg_obss_pd_max_offset:0 \
 			he_bss_color:128 \
-			he_bss_color_enabled:1
+			he_bss_color_enabled:1 \
+			he_ul_mumimo \
+			eht_ulmumimo_80mhz \
+			eht_ulmumimo_160mhz \
+			eht_ulmumimo_320mhz \
+			multiple_bssid \
+			mbssid_group_size \
+			he_6ghz_reg_pwr_type:0
+
+		if [ "$band" = "6g" ]; then
+			append base_cfg "he_6ghz_reg_pwr_type=$he_6ghz_reg_pwr_type" "$N"
+		fi
 
-		he_phy_cap=$(iw phy "$phy" info | sed -n '/HE Iftypes: .*AP/,$p' | awk -F "[()]" '/HE PHY Capabilities/ { print $2 }' | head -1)
+		he_phy_cap=$(eval $sedString | awk -F "[()]" '/HE PHY Capabilities/ { print $2 }' | head -1)
 		he_phy_cap=${he_phy_cap:2}
-		he_mac_cap=$(iw phy "$phy" info | sed -n '/HE Iftypes: .*AP/,$p' | awk -F "[()]" '/HE MAC Capabilities/ { print $2 }' | head -1)
+		he_mac_cap=$(eval $sedString | awk -F "[()]" '/HE MAC Capabilities/ { print $2 }' | head -1)
 		he_mac_cap=${he_mac_cap:2}
 
 		append base_cfg "ieee80211ax=1" "$N"
@@ -472,19 +654,6 @@ mac80211_hostapd_setup_base() {
 			he_spr_psr_enabled:${he_phy_cap:14:2}:0x1:$he_spr_psr_enabled \
 			he_twt_required:${he_mac_cap:0:2}:0x6:$he_twt_required
 
-		if [ "$he_bss_color_enabled" -gt 0 ]; then
-			append base_cfg "he_bss_color=$he_bss_color" "$N"
-			[ "$he_spr_non_srg_obss_pd_max_offset" -gt 0 ] && { \
-				append base_cfg "he_spr_non_srg_obss_pd_max_offset=$he_spr_non_srg_obss_pd_max_offset" "$N"
-				he_spr_sr_control=$((he_spr_sr_control | (1 << 2)))
-			}
-			[ "$he_spr_psr_enabled" -gt 0 ] || he_spr_sr_control=$((he_spr_sr_control | (1 << 0)))
-			append base_cfg "he_spr_sr_control=$he_spr_sr_control" "$N"
-		else
-			append base_cfg "he_bss_color_disabled=1" "$N"
-		fi
-
-
 		append base_cfg "he_default_pe_duration=4" "$N"
 		append base_cfg "he_rts_threshold=1023" "$N"
 		append base_cfg "he_mu_edca_qos_info_param_count=0" "$N"
@@ -511,16 +680,123 @@ mac80211_hostapd_setup_base() {
 		append base_cfg "he_mu_edca_ac_vo_ecwmin=5" "$N"
 		append base_cfg "he_mu_edca_ac_vo_ecwmax=7" "$N"
 		append base_cfg "he_mu_edca_ac_vo_timer=255" "$N"
-	fi
 
-	if [ "$enable_be" != "0" ]; then
-		append base_cfg "ieee80211be=1" "$N"
-		[ "$hwmode" = "a" ] && {
-			append base_cfg "eht_oper_chwidth=$eht_oper_chwidth" "$N"
-			append base_cfg "eht_oper_centr_freq_seg0_idx=$eht_center_seg0" "$N"
-		}
+		if [ -n "$he_ul_mumimo" ]; then
+			if [ "$he_ul_mumimo" -eq 0 ]; then
+				append base_cfg "he_ul_mumimo=0" "$N"
+			elif [ "$he_ul_mumimo" -gt 0 ]; then
+				append base_cfg "he_ul_mumimo=1" "$N"
+			fi
+		else
+			append base_cfg "he_ul_mumimo=-1" "$N"
+		fi
+
+		# If he_bss_color_enabled is set to zero by default, handle enable_color accordingly.
+		# he_bss_color will not work in this case.
+		if [ "$he_bss_color_enabled" -gt 0 ]; then
+			config_get enable_color mac80211 enable_color 1
+			if [ "$enable_color" -eq 1 ]; then
+				bss_color=$(head -1 /dev/urandom | tr -dc '0-9' | head -c2)
+				[ -z "$bss_color" ] && bss_color=0
+				[ "$bss_color" != "0" ] && bss_color=${bss_color#0}
+				bss_color=$((bss_color % 63))
+				bss_color=$((bss_color + 1))
+				append base_cfg "he_bss_color=$bss_color" "$N"
+			fi
+
+			if [ "$he_spr_non_srg_obss_pd_max_offset" -gt 0 ]; then
+				append base_cfg "he_spr_non_srg_obss_pd_max_offset=$he_spr_non_srg_obss_pd_max_offset" "$N"
+				he_spr_sr_control=$((he_spr_sr_control | (1 << 2)))
+			fi
+
+			[ "$he_spr_psr_enabled" -gt 0 ] || he_spr_sr_control=$((he_spr_sr_control | (1 << 0)))
+			append base_cfg "he_spr_sr_control=$he_spr_sr_control" "$N"
+		else
+			append base_cfg "he_bss_color_disabled=1" "$N"
+		fi
+
+		if [ "$enable_be" != "0" ]; then
+			append base_cfg "ieee80211be=1" "$N"
+			[ "$hwmode" = "a" ] && {
+				append base_cfg "eht_oper_chwidth=$eht_oper_chwidth" "$N"
+				append base_cfg "eht_oper_centr_freq_seg0_idx=$eht_center_seg0" "$N"
+			}
+			json_get_vars ru_punct_bitmap:0 ru_punct_ofdma:0 ru_punct_acs_threshold:0 ccfs:0
+			append base_cfg "eht_su_beamformer=1" "$N"
+			append base_cfg "eht_mu_beamformer=1" "$N"
+			append base_cfg "eht_su_beamformee=1" "$N"
+
+			if [ -n "$eht_ulmumimo_80mhz" ]; then
+				if [ "$eht_ulmumimo_80mhz" -eq 0 ]; then
+					append base_cfg "eht_ulmumimo_80mhz=0" "$N"
+				elif [  "$eht_ulmumimo_80mhz" -gt 0 ]; then
+					append base_cfg "eht_ulmumimo_80mhz=1" "$N"
+				fi
+			else
+				append base_cfg "eht_ulmumimo_80mhz=-1" "$N"
+			fi
+
+			if [ -n "$eht_ulmumimo_160mhz" ]; then
+				if [ "$eht_ulmumimo_160mhz" -eq 0 ]; then
+					append base_cfg "eht_ulmumimo_160mhz=0" "$N"
+				elif [  "$eht_ulmumimo_160mhz" -gt 0 ]; then
+					append base_cfg "eht_ulmumimo_160mhz=1" "$N"
+				fi
+			else
+				append base_cfg "eht_ulmumimo_160mhz=-1" "$N"
+			fi
+
+			if [ -n "$eht_ulmumimo_320mhz" ]; then
+				if [ "$eht_ulmumimo_320mhz" -eq 0 ]; then
+					append base_cfg "eht_ulmumimo_320mhz=0" "$N"
+				elif [  "$eht_ulmumimo_320mhz" -gt 0 ]; then
+					append base_cfg "eht_ulmumimo_320mhz=1" "$N"
+				fi
+			else
+				append base_cfg "eht_ulmumimo_320mhz=-1" "$N"
+			fi
+
+			if [ -n "$ru_punct_bitmap" ] && [ "$ru_punct_bitmap" -gt 0 ]; then
+				append base_cfg "ru_punct_bitmap=$ru_punct_bitmap" "$N"
+			fi
+
+			if [ -n "$ru_punct_ofdma" ] && [ "$ru_punct_ofdma" -gt 0 ]; then
+				append base_cfg "ru_punct_ofdma=$ru_punct_ofdma" "$N"
+			fi
+
+			if [ -n "$ru_punct_acs_threshold" ] && [ "$ru_punct_acs_threshold" -gt 0 ]; then
+				append base_cfg "ru_punct_acs_threshold=$ru_punct_acs_threshold" "$N"
+			fi
+
+			[-n "$use_ru_puncture_dfs" ] && append base_cfg "use_ru_puncture_dfs=$use_ru_puncture_dfs" "$N"
+		fi
+
+		if [ "$band" = "6g" ]; then
+			if [ -z "$multiple_bssid" ] && [ "$has_ap" -gt 1 ]; then
+				multiple_bssid=3
+			fi
+		fi
+
+		if [ "$multiple_bssid" == "3" ]; then
+			if [ -z "$mbssid_group_size" ]; then
+				mbssid_group_size=4
+			fi
+		fi
+
+		if [[ "$htmode" == "HE"* ]] || [ "$band" = "6g" ]; then
+			if [ "$has_ap" -gt 1 ]; then
+				append base_cfg "mbssid=$multiple_bssid" "$N"
+			fi
+
+			if [ "$multiple_bssid" == "3" ]; then
+				append base_cfg "mbssid_group_size=$mbssid_group_size" "$N"
+			fi
+		fi
 	fi
 
+	[ -n "$disable_csa_dfs" ] && append base_cfg "disable_csa_dfs=$disable_csa_dfs" "$N"
+	[ -n "$discard_6g_awgn_event" ] && append base_cfg "discard_6g_awgn_event=$discard_6g_awgn_event" "$N"
+
 	hostapd_prepare_device_config "$hostapd_conf_file" nl80211
 	cat >> "$hostapd_conf_file" <<EOF
 ${channel:+channel=$channel}
@@ -535,6 +811,30 @@ EOF
 	json_select ..
 }
 
+mac80211_wds_support_check() {
+	local phy="$1"
+	wds_support=1
+
+	local platform freq board_type
+	platform=$(grep -o "IPQ.*" /proc/device-tree/model | awk -F/ '{print $1}')
+	case "$platform" in
+		"IPQ8074" | "IPQ6018" | "IPQ5018")
+			wds_support=$(cat /sys/module/ath11k/parameters/frame_mode)
+			;;
+		"IPQ9574")
+			freq="$(get_freq "$phy" "$channel" "$band")"
+			board_type=$(grep -o "IPQ.*" /proc/device-tree/model | awk -F/ '{print $3}' | awk -F- '{print $3}')
+
+			if [ "$board_type" == "C6" ] && [ "$freq" -gt 2000 ] && [ "$freq" -lt 3000 ]; then
+				wds_support=$(cat /sys/module/ath11k/parameters/frame_mode)
+			fi
+			;;
+	esac
+
+	echo "$wds_support"
+}
+
+
 mac80211_hostapd_setup_bss() {
 	local phy="$1"
 	local ifname="$2"
@@ -545,17 +845,70 @@ mac80211_hostapd_setup_bss() {
 	append hostapd_cfg "$type=$ifname" "$N"
 
 	hostapd_set_bss_options hostapd_cfg "$phy" "$vif" || return 1
-	json_get_vars wds wds_bridge dtim_period max_listen_int start_disabled
+	json_get_vars wds wds_bridge dtim_period max_listen_int start_disabled ieee80211w beacon_prot ppe_vp
+	json_get_vars unsol_bcast_presp fils_discovery
 
 	set_default wds 0
 	set_default start_disabled 0
 
+	case "$auth_type" in
+		psk|sae|psk-sae|owe|eap*|wep|sae-mixed|ft-sae-ext-key)
+			if [ "$ieee80211w" -gt 0 ] && [ "$beacon_prot" -gt 0 ]; then
+				append hostapd_cfg "beacon_prot=1" "$N"
+			fi
+		;;
+	esac
+
 	[ "$wds" -gt 0 ] && {
+		wds_support=$(mac80211_wds_support_check "$phy")
+                if [ "$wds_support" -ne 1 ]; then
+                        echo WDS is supported only in native wifi mode for ath11k driver. Kindly update the config > /dev/ttyMSM0
+                        return
+                fi
+
 		append hostapd_cfg "wds_sta=1" "$N"
 		[ -n "$wds_bridge" ] && append hostapd_cfg "wds_bridge=$wds_bridge" "$N"
 	}
 	[ "$staidx" -gt 0 -o "$start_disabled" -eq 1 ] && append hostapd_cfg "start_disabled=1" "$N"
 
+	if [ "$band" = "6g" ]; then
+		fils_cfg=
+		if [ "$unsol_bcast_presp" -gt 0 ] && [ "$unsol_bcast_presp" -le 20 ]; then
+			append fils_cfg "unsol_bcast_probe_resp_interval=$unsol_bcast_presp" "$N"
+		elif [ "$fils_discovery" -gt 0 ] && [ "$fils_discovery" -le 20 ]; then
+			append fils_cfg "fils_discovery_max_interval=$fils_discovery" "$N"
+		else
+			append fils_cfg "fils_discovery_max_interval=20" "$N"
+		fi
+
+		if [ -n "$multiple_bssid" ] && [ "$multiple_bssid" -ge 1 ] && [ "$type" == "interface" ]; then
+			append hostapd_cfg "$fils_cfg" "$N"
+		elif [ -z "$multiple_bssid" ] || [ "$multiple_bssid" -eq 0 ]; then
+			append hostapd_cfg "$fils_cfg" "$N"
+		elif [ -n "$multiple_bssid" ] && [ "$multiple_bssid" -eq 3 ] && [ $((id % mbssid_group_size)) == "0" ]; then
+			append hostapd_cfg "$fils_cfg" "$N"
+		fi
+        fi
+
+	if [[ "$htmode" == "EHT"* ]]; then
+		append hostapd_cfg "mld_ap=1" "$N"
+	fi
+
+	case "$ppe_vp" in
+		"passive")
+			append hostapd_cfg "ppe_vp=1" "$N"
+			;;
+		"active")
+			append hostapd_cfg "ppe_vp=2" "$N"
+			;;
+		"ds")
+			append hostapd_cfg "ppe_vp=3" "$N"
+			;;
+		*)
+			append hostapd_cfg "ppe_vp=3" "$N"
+			;;
+	esac
+
 	cat >> /var/run/hostapd-$phy$vif_phy_suffix.conf <<EOF
 $hostapd_cfg
 bssid=$macaddr
@@ -656,7 +1009,42 @@ find_phy() {
 }
 
 mac80211_check_ap() {
-	has_ap=1
+        has_ap=$((has_ap+1))
+}
+
+mac80211_get_band_name() {
+	local radio_id=$1
+
+	[ "$radio_id" = "-1" ] && return $band
+
+	freq_range=$(iw ${phy} info | grep -A 1 "Idx $radio_id:" | grep "Frequency Range:" | awk '{print $3, $6}')
+	set -- $freq_range
+	freq1=$1
+	freq2=$2
+
+	if [ "$freq1" -gt 2400 ] && [ "$freq2" -lt 2495 ]; then
+		band_name=2g
+	elif [ "$freq1" -gt 5100 ] && [ "$freq2" -lt 5900 ]; then
+		if [ "$freq1" -gt 5100 ] && [ "$freq2" -lt 5400 ]; then
+			band_name=5gl
+		elif [ "$freq1" -gt 5400 ] && [ "$freq2" -lt 5900 ]; then
+			band_name=5gh
+		else
+			band_name=5g
+		fi
+	elif [ "$freq1" -gt 5900 ] && [ "$freq2" -lt 7200 ]; then
+		if [ "$freq1" -gt 5900 ] && [ "$freq2" -lt 6425 ]; then
+			band_name=6gl
+		elif [ "$freq1" -gt 6500 ] && [ "$freq2" -lt 7200 ]; then
+			band_name=6gh
+		else
+			band_name=6g
+		fi
+	else
+		band_name=invalid
+	fi
+
+	echo $band_name
 }
 
 mac80211_set_ifname() {
@@ -666,9 +1054,10 @@ mac80211_set_ifname() {
 }
 
 mac80211_prepare_vif() {
+	ppe_vp="ds"
 	json_select config
 
-	json_get_vars ifname mode ssid wds powersave macaddr enable wpa_psk_file vlan_file
+	json_get_vars ifname mode ssid wds powersave macaddr enable wpa_psk_file vlan_file ppe_vp mld
 
 	[ -n "$ifname" ] || {
 		local prefix;
@@ -679,19 +1068,41 @@ mac80211_prepare_vif() {
 		monitor) prefix=mon;;
 		esac
 
-		mac80211_set_ifname "$phy$vif_phy_suffix" "$prefix"
+                if [ "$is_wiphy_multi_radio" -eq 1 ]; then
+                        if [[ "$htmode" == EHT* ]] && [ -n "$mld" ]; then
+				config_get mld_ifname "$mld" ifname
+				if [ -z "$mld_ifname" ]; then
+					ifname=$phy-$mld
+				else
+					ifname="$mld_ifname"
+				fi
+                        else
+                                mac80211_set_ifname "$phy-$band_name" "$prefix"
+                        fi
+                else
+                        mac80211_set_ifname "$phy$vif_phy_suffix" "$prefix"
+                fi
+
 	}
 
+	[ -z $ppe_vp ] && ppe_vp="ds"
+
+        if [ $mode == "mesh" ] && [ $ppe_vp == "ds" ]; then
+                ppe_vp="passive"
+        fi
+
+
 	append active_ifnames "$ifname"
 	set_default wds 0
 	set_default powersave 0
 	json_add_string _ifname "$ifname"
+	json_add_string _ppe_vp "$ppe_vp"
 
 	default_macaddr=
 	if [ -z "$macaddr" ]; then
 		macaddr="$(mac80211_generate_mac $phy)"
 		macidx="$(($macidx + 1))"
-		default_macaddr=1
+		#default_macaddr=1
 	elif [ "$macaddr" = 'random' ]; then
 		macaddr="$(macaddr_random)"
 	fi
@@ -699,11 +1110,13 @@ mac80211_prepare_vif() {
 	json_add_string _default_macaddr "$default_macaddr"
 	json_select ..
 
-
-	[ "$mode" == "ap" ] && {
-		[ -z "$wpa_psk_file" ] && hostapd_set_psk "$ifname"
-		[ -z "$vlan_file" ] && hostapd_set_vlan "$ifname"
-	}
+	[ -f /tmp/mlo_support.txt ] && mlo_add_flag=$(cat /tmp/mlo_support.txt)
+	if [ $mlo_add_flag -eq 0 ]; then
+		[ "$mode" == "ap" ] && {
+			[ -z "$wpa_psk_file" ] && hostapd_set_psk "$ifname"
+			[ -z "$vlan_file" ] && hostapd_set_vlan "$ifname"
+		}
+	fi
 
 	json_select config
 
@@ -852,6 +1265,7 @@ mac80211_setup_mesh() {
 	json_add_string htmode "$iw_htmode"
 	[ -n "$mcval" ] && json_add_string mcast-rate "$mcval"
 	json_add_int beacon-interval "$beacon_int"
+	json_add_string ru-puncturing-bitmap "$ru_punct_bitmap"
 	mac80211_add_mesh_params
 
 	json_close_object
@@ -877,6 +1291,7 @@ mac80211_set_vif_txpower() {
 
 	json_select config
 	json_get_var ifname _ifname
+	json_get_var ppe_vp _ppe_vp
 	json_get_vars vif_txpower
 	json_select ..
 
@@ -886,6 +1301,25 @@ mac80211_set_vif_txpower() {
 	else
 		iw dev "$ifname" set txpower auto
 	fi
+
+	iw dev "$ifname" set_intf_offload type "$ppe_vp"
+}
+
+mac80211_set_fq_limit() {
+        json_select data
+        json_get_vars ifname
+        json_select ..
+
+        json_select config
+        json_get_vars fq_limit
+
+        if [ "$fq_limit" -gt 0 ]; then
+                tc qdisc add dev "$ifname" parent :1 fq_codel limit "$fq_limit"
+                tc qdisc add dev "$ifname" parent :2 fq_codel limit "$fq_limit"
+                tc qdisc add dev "$ifname" parent :3 fq_codel limit "$fq_limit"
+                tc qdisc add dev "$ifname" parent :4 fq_codel limit "$fq_limit"
+        fi
+        json_select ..
 }
 
 wpa_supplicant_init_config() {
@@ -910,6 +1344,13 @@ wpa_supplicant_add_interface() {
 	json_add_string ctrl "$_rpath"
 	json_add_string iface "$ifname"
 	json_add_string mode "$mode"
+	json_add_int radio "$radio"
+	if [ "$mode" = "sta" ]; then
+		[ -n "$mld" ] && {
+			json_add_string mld "$mld"
+		}
+	fi
+
 	json_add_string config "$_config"
 	[ -n "$default_macaddr" ] || json_add_string macaddr "$macaddr"
 	[ -n "$network_bridge" ] && json_add_string bridge "$network_bridge"
@@ -934,6 +1375,7 @@ wpa_supplicant_set_config() {
 	json_add_int radio "$radio"
 	json_add_int num_global_macaddr "$num_global_macaddr"
 	json_add_boolean defer 1
+	[ -n "$mld" ] && json_add_boolean is_ml 1
 	local data="$(json_dump)"
 
 	json_cleanup
@@ -950,7 +1392,7 @@ wpa_supplicant_set_config() {
 	[ "$ret" != 0 -o -z "$supplicant_res" ] && wireless_setup_vif_failed WPA_SUPPLICANT_FAILED
 
 	wireless_add_process "$(jsonfilter -s "$supplicant_res" -l 1 -e @.pid)" "/usr/sbin/wpa_supplicant" 1 1
-
+	update_primary_link
 }
 
 hostapd_set_config() {
@@ -976,10 +1418,12 @@ hostapd_set_config() {
 wpa_supplicant_start() {
 	local phy="$1"
 	local radio="$2"
+	local is_mld="false"
 
 	[ -n "$wpa_supp_init" ] || return 0
+	[ -n "$mld" ] && is_mld="true"
 
-	ubus_call wpa_supplicant config_set '{ "phy": "'"$phy"'", "radio": '"$radio"', "num_global_macaddr": '"$num_global_macaddr"' }' > /dev/null
+	ubus_call wpa_supplicant config_set '{ "phy": "'"$phy"'", "radio": '"$radio"', "num_global_macaddr": '"$num_global_macaddr"', "is_ml": '"$is_mld"' }' > /dev/null
 }
 
 mac80211_setup_supplicant() {
@@ -991,7 +1435,7 @@ mac80211_setup_supplicant() {
 	if [ "$mode" = "sta" ]; then
 		wpa_supplicant_add_network "$ifname"
 	else
-		wpa_supplicant_add_network "$ifname" "$freq" "$htmode" "$hostapd_noscan"
+		wpa_supplicant_add_network "$ifname" "$freq" "$htmode" "$hostapd_noscan" "$ru_punct_bitmap" "$disable_csa_dfs" "$ccfs"
 	fi
 
 	wpa_supplicant_add_interface "$ifname" "$mode"
@@ -1007,13 +1451,14 @@ mac80211_setup_vif() {
 	json_get_var ifname _ifname
 	json_get_var macaddr _macaddr
 	json_get_var default_macaddr _default_macaddr
-	json_get_vars mode wds powersave
+	json_get_vars mode wds powersave mld
 
 	set_default powersave 0
 	set_default wds 0
 
 	case "$mode" in
 		mesh)
+			freq_list=$(get_sta_freq_list "$phy" "$radio" "$band_name")
 			json_get_vars $MP_CONFIG_INT $MP_CONFIG_BOOL $MP_CONFIG_STRING
 			wireless_vif_parse_encryption
 			[ -z "$htmode" ] && htmode="NOHT";
@@ -1032,6 +1477,17 @@ mac80211_setup_vif() {
 			fi
 		;;
 		sta)
+			if [ -n "$mld" ]; then
+				for value in $_sta_radios
+				do
+					bname=$(mac80211_get_band_name $value)
+					flist=$(get_sta_freq_list "$phy" "$value" "$bname")
+					append freq_list "$flist"
+				done
+				freq_list=$freq_list
+			else
+				freq_list=$(get_sta_freq_list "$phy" "$radio" "$band_name")
+			fi
 			mac80211_setup_supplicant || failed=1
 		;;
 		monitor)
@@ -1091,6 +1547,9 @@ mac80211_reset_config() {
 }
 
 mac80211_set_suffix() {
+
+	band_name=$(mac80211_get_band_name $radio)
+
 	[ "$radio" = "-1" ] && radio=
 	phy_suffix="${radio:+:$radio}"
 	vif_phy_suffix="${radio:+.$radio}"
@@ -1098,19 +1557,33 @@ mac80211_set_suffix() {
 }
 
 drv_mac80211_setup() {
+	local device=$1
+
+	mac80211_derive_ml_info
+
 	json_select config
 	json_get_vars \
 		radio phy macaddr path \
-		country chanbw distance \
+		country chanbw distance band\
 		txpower \
-		rxantenna txantenna \
+		rxantenna txantenna antenna_gain\
 		frag rts beacon_int:100 htmode \
-		num_global_macaddr:1 multiple_bssid
+		num_global_macaddr:1 multiple_bssid \
+		eht_ulmumimo_80mhz eht_ulmumimo_160mhz eht_ulmumimo_320mhz \
+		ccfs disable_csa_dfs ru_punct_bitmap
 	json_get_values basic_rate_list basic_rate
 	json_get_values scan_list scan_list
 	json_select ..
 
-	mac80211_set_suffix
+	if [ ${#device} -eq 12 ]; then
+		is_wiphy_multi_radio=1
+	fi
+
+	if [ "$is_wiphy_multi_radio" -eq 1 ]; then
+		echo [debug] $device - radio as $radio > /tmp/logs
+        else
+                radio=-1
+        fi
 
 	json_select data && {
 		json_get_var prev_rxantenna rxantenna
@@ -1120,50 +1593,80 @@ drv_mac80211_setup() {
 
 	find_phy || {
 		echo "Could not find PHY for device '$1'"
-		wireless_set_retry 0
+		wireless_set_retry 1
 		return 1
 	}
 
+	mac80211_set_suffix
+
+	[ -f /tmp/mlo_support.txt ] && mlo_add_flag=$(cat /tmp/mlo_support.txt)
+	if [ $mlo_add_flag -eq 0 ]; then
+		if [ "$(cat /sys/module/ath12k/parameters/ppe_rfs_support)" == 'Y' ]; then
+			# Note: ppe_vp_accel and ppe_vp_rfs are mutually exclusive.
+			#       ppe_vp_accel enables PPE acceleration path and ppe_vp_rfs
+			#       is expected to enable only flow steering for VLAN type
+			#       interface (eg: WDS root).
+			echo 1 >> /sys/module/mac80211/parameters/ppe_vp_rfs
+			# Note: Format is default MLO mask followed by band specific core masks
+			#       in order of 2 GHz, 5 GHz and 6GHz bands
+			#       echo <DEFAULT/ MLO MASK>,<2GHZ MASK>,<5GHZ MASK>,<6GHZ_MASK>
+			echo 0x7,0x7,0x7,0x7 > /sys/module/ath12k/parameters/rfs_core_mask
+
+			if [ "$(cat /sys/module/mac80211/parameters/ppe_vp_accel)" == 'Y' ]; then
+				echo "ppe_vp_accel is enabled. Please disable to support RFS on WDS" > /dev/ttyMSM0
+			fi
+
+			if echo "$(cat /sys/sfe/ppe_rfs_feature)" | grep -q "disabled"; then
+				echo 1 >> /sys/sfe/ppe_rfs_feature
+				echo "enabled ppe_rfs_feature" > /dev/ttyMSM0
+			fi
+		fi
+	fi
+
 	local wdev
 	local cwdev
 	local found
 
 	# convert channel to frequency
-	[ "$auto_channel" -gt 0 ] || freq="$(get_freq "$phy" "$channel" "$band")"
+	if [ $mlo_add_flag -eq 0 ]; then
+		[ "$auto_channel" -gt 0 ] || freq="$(get_freq "$phy" "$channel" "$band")"
 
-	[ -n "$country" ] && {
-		iw reg get | grep -q "^country $country:" || {
-			iw reg set "$country"
-			sleep 1
+		[ -n "$country" ] && {
+			iw reg get | grep -q "^country $country:" || {
+				iw reg set "$country"
+				sleep 1
+			}
 		}
-	}
+	fi
 
 	hostapd_conf_file="/var/run/hostapd-$phy$vif_phy_suffix.conf"
 
 	macidx=0
 	staidx=0
 
-	[ -n "$chanbw" ] && {
-		for file in /sys/kernel/debug/ieee80211/$phy/ath9k*/chanbw /sys/kernel/debug/ieee80211/$phy/ath5k/bwmode; do
-			[ -f "$file" ] && echo "$chanbw" > "$file"
-		done
-	}
+	if [ $mlo_add_flag -eq 0 ]; then
+		[ -n "$chanbw" ] && {
+			for file in /sys/kernel/debug/ieee80211/$phy/ath9k*/chanbw /sys/kernel/debug/ieee80211/$phy/ath5k/bwmode; do
+				[ -f "$file" ] && echo "$chanbw" > "$file"
+			done
+		}
 
-	set_default rxantenna 0xffffffff
-	set_default txantenna 0xffffffff
-	set_default distance 0
+		set_default rxantenna 0xffffffff
+		set_default txantenna 0xffffffff
+		set_default distance 0
 
-	[ "$txantenna" = "all" ] && txantenna=0xffffffff
-	[ "$rxantenna" = "all" ] && rxantenna=0xffffffff
+		[ "$txantenna" = "all" ] && txantenna=0xffffffff
+		[ "$rxantenna" = "all" ] && rxantenna=0xffffffff
 
-	[ "$rxantenna" = "$prev_rxantenna" -a "$txantenna" = "$prev_txantenna" ] || mac80211_reset_config "$phy"
-	wireless_set_data phy="$phy" radio="$radio" txantenna="$txantenna" rxantenna="$rxantenna"
+		[ "$rxantenna" = "$prev_rxantenna" -a "$txantenna" = "$prev_txantenna" ] || mac80211_reset_config "$phy"
+		wireless_set_data phy="$phy" radio="$radio" txantenna="$txantenna" rxantenna="$rxantenna"
 
-	iw phy "$phy" set antenna $txantenna $rxantenna >/dev/null 2>&1
-	iw phy "$phy" set distance "$distance" >/dev/null 2>&1
+		iw phy "$phy" set antenna $txantenna $rxantenna >/dev/null 2>&1
+		iw phy "$phy" set distance "$distance" >/dev/null 2>&1
 
-	[ -n "$frag" ] && iw phy "$phy" set frag "${frag%%.*}"
-	[ -n "$rts" ] && iw phy "$phy" set rts "${rts%%.*}"
+		[ -n "$frag" ] && iw phy "$phy" set frag "${frag%%.*}"
+		[ -n "$rts" ] && iw phy "$phy" set rts "${rts%%.*}"
+	fi
 
 	has_ap=
 	hostapd_ctrl=
@@ -1177,6 +1680,11 @@ drv_mac80211_setup() {
 	for_each_interface "sta adhoc mesh" mac80211_set_noscan
 	[ -n "$has_ap" ] && mac80211_hostapd_setup_base "$phy"
 
+        if [ "$mlo_add_flag" = 1 ]; then
+		for_each_interface "ap" mac80211_prepare_vif
+                return;
+        fi
+
 	local prev
 	json_set_namespace wdev_uc prev
 	json_init
@@ -1199,6 +1707,24 @@ drv_mac80211_setup() {
 	json_set_namespace "$prev"
 
 	for_each_interface "ap sta adhoc mesh monitor" mac80211_set_vif_txpower
+
+	config_get enable_smp_affinity mac80211 enable_smp_affinity 0
+	if [ "$enable_smp_affinity" -eq 1 ]; then
+		[ -f "/lib/smp_affinity_settings.sh" ] && {
+			. /lib/smp_affinity_settings.sh
+			enable_smp_affinity_wifi
+		}
+		[ -f "/lib/update_smp_affinity.sh" ] && {
+			. /lib/update_smp_affinity.sh
+			enable_smp_affinity_wigig
+		}
+	fi
+
+	[ -f "/lib/performance.sh" ] && {
+		. /lib/performance.sh
+	}
+
+	for_each_interface "ap mesh" mac80211_set_fq_limit
 	wireless_set_up
 }
 
@@ -1233,4 +1759,80 @@ drv_mac80211_teardown() {
 	mac80211_reset_config "$phy"
 }
 
+update_primary_link()
+{
+        if [ -n "$mld_names" ]; then
+                for mld_iface in $mld_names; do
+                        config_get mld_primary_link "$mld_iface" primary_link
+                        config_get mld_ifname "$mld_iface" ifname
+                        if [ -n "$mld_primary_link" ]; then
+                                while true;
+                                do
+                                        ifname_state="$(hostapd_cli -i "$mld_ifname" status 2> /dev/null | grep state | cut -d'=' -f 2)"
+                                        if [ "$ifname_state" = "ENABLED" ]; then
+                                                echo "$mld_primary_link" > /sys/kernel/debug/ieee80211/phy"${phy#phy}"/netdev:"$mld_ifname"/primary_link
+                                                break;
+                                        fi
+                                done
+                        fi
+                done
+        fi
+
+}
+
+
+_sta_radios=
+mac80211_derive_ml_info() {
+	local _mlds
+	local sta_mldevices
+	config_load wireless
+
+	mac80211_get_wifi_mlds() {
+		append _mlds "$1"
+	}
+	config_foreach mac80211_get_wifi_mlds wifi-mld
+
+	if [ -z "$_mlds" ]; then
+		return
+	fi
+
+	mac80211_get_wifi_ifaces() {
+	        config_get iface_mode "$1" mode
+	        if [ -n "$iface_mode" ] && [ "$iface_mode" = "sta" ]; then
+	                 append _staifaces "$1"
+	        fi
+	}
+	config_foreach mac80211_get_wifi_ifaces wifi-iface
+
+	if [ -z "$_staifaces" ]; then
+	        return
+	fi
+
+	for _mld in $_mlds
+	do
+	        for _staifname in $_staifaces
+	        do
+	                config_get mld_name "$_staifname" mld
+	                config_get mldevice "$_staifname" device
+	                if [ ${#mldevice} -ne 12 ]; then
+	                        continue;
+	                fi
+
+	                if ! [[ "$sta_mldevices" =~ "$mldevice" ]]; then
+	                        append sta_mldevices "$mldevice"
+
+	                        if [ -n "$mld_name" ] &&  [ "$_mld" = "$mld_name" ]; then
+	                                config_get disabled "$mldevice" disabled
+	                                if [ "$disabled" -eq 1 ]; then
+	                                        continue;
+	                                fi
+	                                config_get radio "$mldevice" radio
+	                                append _sta_radios $radio
+	                        fi
+	                fi
+	        done
+	done
+}
+
+
 add_driver mac80211
-- 
2.34.1

